db/src/db.rs

```diff
@@ -870,6 +889,90 @@ pub fn update_partition_map(conn: &rusqlite::Connection, partition_map: &Partiti
         .chain_err(|| "Could not update partition map")
 }
 
+/// Update the metadata materialized views based on the given metadata report.
+///
+/// This updates the "entids", "idents", and "schema" materialized views, copying directly from the
+/// "datoms" and "transactions" table as appropriate.
+pub fn update_metadata(conn: &rusqlite::Connection, _old_schema: &Schema, new_schema: &Schema, metadata_report: &metadata::MetadataReport) -> Result<()>
+{
+    use metadata::AttributeAlteration::*;
+
+    // Populate the materialized view directly from datoms and transactions.  This might generalize
+    // nicely as we expand the set of materialized views.
+    // TODO: consider doing this in fewer SQLite execute() invocations.
+    // TODO: use concat! to avoid creating String instances.
+    if !metadata_report.idents_altered.is_empty() {
+        // Entids is the materialized view of the [entid :db/ident ident] slice of datoms.
+        conn.execute(format!("DELETE FROM entids").as_str(),
+                     &[])?;
+        conn.execute(format!("INSERT INTO entids SELECT e, a, v, value_type_tag FROM datoms WHERE a IN {}", entids::IDENTS_SET.as_str()).as_str(),
+                     &[])?;
+
+        // Idents is the materialized view of the [entid :db/ident ident _ added] slice of
+        // transactions.  This keeps outdated idents around, just like Datomic does.
+        conn.execute(format!("DELETE FROM idents").as_str(),
+                     &[])?;
+        // Take the last, as determined by tx, added [e :db/ident v _ true].  This will pick up
+        // :db/idents that have been retracted and/or re-purposed.
+        conn.execute(format!("INSERT INTO idents SELECT e, a, v, value_type_tag FROM (SELECT e, a, v, value_type_tag, MAX(tx) FROM transactions WHERE a IN {} AND added = 1 GROUP BY e, a, v, value_type_tag, tx)", entids::IDENTS_SET.as_str()).as_str(),
+                     &[])?;
+    }
+
+    for &entid in &metadata_report.attributes_installed {
+        conn.execute(format!("INSERT INTO schema SELECT e, a, v, value_type_tag FROM datoms WHERE e = ? AND a IN {}", entids::SCHEMA_SET.as_str()).as_str(),
```

Comment:
In general, I'm for cached statements (especially since Rusqlite makes them easy to use) but I intentionally didn't optimize this process.  I expect it to be infrequent ... oh, prepared vs. cached.  Prepared I will do!  This is not efficient at this time because walking `transactions` in this way is very expensive but it's easy to improve and I'll file follow-up to track that.